contributor-docs/adrs/adr-020-live-regions.md

```diff
@@ -0,0 +1,165 @@
+This package will be included and published from the `primer/react` repo.
+
+The custom element exposes a way to make announcements that can be used
+across frameworks. This makes it useful not only for Primer but GitHub as a
+whole.
+
+In addition, `@primer/react` will leverage and export the following helpers for
+use within Primer React and GitHub:
+
+- `announce()` and `announceFromElement()`, programmatic helpers for making
+  announcements
+- The `Status` component to correspond with `role="status"`
+- The `Alert` component to correspond with `role="alert"`
+
+Within `@primer/react`, we should lint against usage of `aria-live` and the
+corresponding roles (if possible) and suggest using these alternatives instead.
+
+### Impact
+
+This decision will impact existing usage of `aria-live`, `role="status"`, and
+`role="alert"` for the components listed below. These components will need to be
+updated to use the new approach, using one of the following approaches:
+
+- Use `announce()` or `announceFromElement()`
+- Use `Status` or `Alert`
+
+In addition, we should make sure that `<live-region>` is successfully included
+in GitHub.
+
+#### Instances of `aria-live`
+
+- InlineAutocomplete
+- InputValidation
+- SelectPanel
+- TreeView
+
+### Instances of `role="status"`
+
+- LiveRegion
+- TreeView
+- ActionList examples
+
+### Instances of `role="alert"`
+
+None
+
+### Instances of `LiveRegion`
+
+- DataTable
+
+## Examples
+
+### Announce on page load
+
+```tsx
+import {Status} from '@primer/react'
+
+function ExampleComponent() {
+  return <Status>Example page load message</Status>
+}
+```
+
+### Announce on content change
+
+```tsx
+import {Status} from '@primer/react'
+import {useState} from 'react'
+
+function ExampleComponent() {
+  const [count, setCount] = useState(0)
+  return (
+    <button
+      type="button"
+      onClick={() => {
+        setCount(count + 1)
+      }}
+    >
+      <Status>Count {count}</Status>
+    </button>
+  )
+}
+```
+
+### Announce programmatically
+
+```tsx
+import {announce} from '@primer/react'
+import {useState} from 'react'
+
+function ExampleComponent() {
+  const [results, setResults] = useState(data)
+
+  return (
+    <>
+      <input
+        type="text"
+        onChange={event => {
+          const filteredResults = data.filter(item => {
+            /* ... */
+          })
+          setResults(filteredResults)
+          announce(`${filteredResults.length} results available`)
+        }}
+      />
+      {/* ... */}
+    </>
+  )
+}
+```
+
+## Alternatives
+
+## Unresolved questions
+
+- What should happen if a component makes excessive announcements due to a
+  component that it renders? Is it possible to "group" announcements?
+- For programmatic announcements, how do we interop with existing proposals in
+  this space for `ariaNotify`?
+- When should something use an ARIA Live region role component versus just using
```

Comment:
Maybe when the region exists in a `Dialog`. This is because, in most screen readers / browsers, the live region wont announce unless it's directly inside of the dialog. Since our current utility doesn't handle this, I'd probably recommend using a live region directly. The only tricky thing about this is, if the live region doesn't exist at page load, there's the chance it won't announce.

Reply:
@TylerJDev one nice thing about `announce()` or `announceFromElement()` is that they accept a `from` argument that will try to find a live region and create one if it doesn't exist. The logic for this is over at: https://github.com/primer/live-region-element/blob/main/packages/live-region-element/src/query.ts#L3 so that if the `from` element is within a  `dialog` the live region will attach itself to that dialog. `Status` and `Alert` will also do this automatically if one doesn't exist.

Either way, the way these utilities work is that they will make sure a live region exists _before_ making the announcement and hopefully that helps improve the robustness here 🤞